[Thumb] Resolve FIXME: Use 'mov hi, $src; mov $dst, hi'

AZero13 opened the pull request:
Consider the following:

        ldr     r0, [r4]
        ldr     r7, [r0, #4]
        cmp     r7, r3
        bhi     .LBB0_6
        cmp     r0, r2
        push    {r0}
        pop     {r4}
        bne     .LBB0_3
        movs    r0, r6
        pop     {r4, r5, r6, r7}
        pop     {r1}
        bx      r1

Here is a snippet of the generated THUMB1 code of the K&R malloc function that clang currently compiles to.

push    {r0} ends up being popped to pop {r4}.

movs r4, r0 would destroy the flags set by cmp right above.

The compiler has no alternative in this case, except one:
the only alternative is to transfer through a high register.

However, it seems like LLVM does not consider that this is a valid approach, even though it is a free clobbering a high register.

This patch addresses the FIXME so the compiler can do that when it can in r10 or r11, or r12.

AtariDreams commented:
@nikic

AtariDreams commented:
Full source if anyone is curious:

https://godbolt.org/z/h7Pq3s6dM

efriedma-quic commented:
(See https://lists.llvm.org/pipermail/llvm-dev/2014-August/075967.html for additional context.)

r10 and r11 are callee-save, so you'd need to make sure they get spilled.  We have code to do the spilling, but it's not triggering here. I guess this code runs too late for that to trigger?

AtariDreams commented:
> (See https://lists.llvm.org/pipermail/llvm-dev/2014-August/075967.html for additional context.)
> 
> r10 and r11 are callee-save, so you'd need to make sure they get spilled. We have code to do the spilling, but it's not triggering here. I guess this code runs too late for that to trigger?

Pretty sure only r12 is, hence why this is not an option.

AtariDreams commented:
> (See https://lists.llvm.org/pipermail/llvm-dev/2014-August/075967.html for additional context.)
> 
> r10 and r11 are callee-save, so you'd need to make sure they get spilled. We have code to do the spilling, but it's not triggering here. I guess this code runs too late for that to trigger?

r10 - Caller save (in ARM mode, unused in Thumb mode)
r11 Caller save (in ARM mode, unused in Thumb mode)

AtariDreams commented:
> r10 and r11 are callee-save, so you'd need to make sure they get spilled

r10 and r11 are callee-save, so you'd need to make sure they get spilled

I only think this is an issue if the compiler were to randomly switch between thumb and arm mode and vice versa mid function, which is a thing that only really happens in handwritten code anyway.

efriedma-quic commented:
Pretty sure that's just wrong (or referring to some other calling convention).  From the specification (https://github.com/ARM-software/abi-aa/releases):

> A subroutine must preserve the contents of the registers r4-r8, r10, r11 and SP (and r9 in PCS variants that designate r9 as v6).

AtariDreams commented:
@efriedma-quic Okay, I fixed the issues!

efriedma-quic commented:
For r10 and r11, as the tests show, they're almost never available; the compiler won't naturally spill them.  You can force them to spill using inline asm, but that's so rare it's not really relevant.  So I don't really see the point of checking, unless you're planning some sort of followup.

For r12... this is probably okay?  We don't use r12 in the scavenger anymore (see ab1d73ee).  We still use R12 as a scratch register in a few different places, but not in a way that would conflict, I think.  It would be helpful if you could take some time to verify that.

If there's some way r12 can actually be live (inline asm?), we should make sure there's at least one test that actually still exercises the code to generate push/pop.

AtariDreams commented:
> For r10 and r11, as the tests show, they're almost never available; the compiler won't naturally spill them. You can force them to spill using inline asm, but that's so rare it's not really relevant. So I don't really see the point of checking, unless you're planning some sort of followup.
> 
> For r12... this is probably okay? We don't use r12 in the scavenger anymore (see [ab1d73e](https://github.com/llvm/llvm-project/commit/ab1d73ee32481e3033ebbe5a68afafb24da3df2e)). We still use R12 as a scratch register in a few different places, but not in a way that would conflict, I think. It would be helpful if you could take some time to verify that.
> 
> If there's some way r12 can actually be live (inline asm?), we should make sure there's at least one test that actually still exercises the code to generate push/pop.

I didn't know if the compiler for some reason ever did passes that would move instructions around and interleave multiple copies, so I did that just in case. But honestly I think this is a better way:

AtariDreams commented:
> For r10 and r11, as the tests show, they're almost never available; the compiler won't naturally spill them. You can force them to spill using inline asm, but that's so rare it's not really relevant. So I don't really see the point of checking, unless you're planning some sort of followup.
> 
> For r12... this is probably okay? We don't use r12 in the scavenger anymore (see [ab1d73e](https://github.com/llvm/llvm-project/commit/ab1d73ee32481e3033ebbe5a68afafb24da3df2e)). We still use R12 as a scratch register in a few different places, but not in a way that would conflict, I think. It would be helpful if you could take some time to verify that.
> 
> If there's some way r12 can actually be live (inline asm?), we should make sure there's at least one test that actually still exercises the code to generate push/pop.

I think it is best to delegate this to find the first high register available rather than always go with r12 though maybe I should do r12 always

AtariDreams commented:
@efriedma-quic Thoughts?

AtariDreams commented:
@topperc @efriedma-quic Thoughts?

AtariDreams commented:
> For r10 and r11, as the tests show, they're almost never available; the compiler won't naturally spill them. You can force them to spill using inline asm, but that's so rare it's not really relevant. So I don't really see the point of checking, unless you're planning some sort of followup.
> 
> For r12... this is probably okay? We don't use r12 in the scavenger anymore (see [ab1d73e](https://github.com/llvm/llvm-project/commit/ab1d73ee32481e3033ebbe5a68afafb24da3df2e)). We still use R12 as a scratch register in a few different places, but not in a way that would conflict, I think. It would be helpful if you could take some time to verify that.
> 
> If there's some way r12 can actually be live (inline asm?), we should make sure there's at least one test that actually still exercises the code to generate push/pop.

If someone is using r12 in inline asm and then not using it, they should not be surprised when the compiler does not know what to do and clobbers it anyway, because this would happen with any other register. those who write inline asm manually specifying r12 surely has a plan to use r12 in some way or another that is dictated by the asm. it is not up to the compiler to guess what the asm is doing anyway.

AZero13 committed:
[Thumb] Resolve FIXME: Use 'mov hi, $src; mov $dst, hi'

Consider the following:

        ldr     r0, [r4]
        ldr     r7, [r0, #4]
        cmp     r7, r3
        bhi     .LBB0_6
        cmp     r0, r2
        push    {r0}
        pop     {r4}
        bne     .LBB0_3
        movs    r0, r6
        pop     {r4, r5, r6, r7}
        pop     {r1}
        bx      r1

Here is a snippet of the generated THUMB1 code of the K&R malloc function that clang currently compiles to.

push    {r0} ends up being popped to pop {r4}.

movs r4, r0 would destroy the flags set by cmp right above.

The compiler has no alternative in this case, except one:
the only alternative is to transfer through a high register.

However, it seems like LLVM does not consider that this is a valid approach, even though it is a free clobbering a high register.

This patch addresses the FIXME so the compiler can do that when it can in r10 or r11, or r12.

AtariDreams commented:
@efriedma-quic Ping?

AtariDreams commented:
@topperc Thoughts?

TNorthover reviewed (commented):
Mostly looks pretty sensible. Couple of comments.

AtariDreams commented:
@TNorthover Addressed!

TNorthover reviewed (approved):
Thanks. I think this looks good now.

AtariDreams commented:
@TNorthover Failure looks unrelated
